PEP 705: TypedMapping (#2997)
python/peps: pep-0705.rst

    from typing import NotRequired, TypedMapping

    class Movie(TypedMapping):
        name: str
        year: NotRequired[int | None]

In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.

Specification
=============

The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:

* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``
* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
* Subclasses can narrow field types, in the same manner as other protocols

> All current and future features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.

Reviewer: ```suggestion
All other current features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

* Make @JelleZijlstra 's [comment](https://github.com/python/peps/pull/2997#discussion_r1096703542) removing "and future" into a suggestion
* Also explicitly clarify that it is all features except for those listed above

Reviewer: This was obsoleted and de-facto implemented by further changes

    b: B = {'x': 0, 'y': 'foo'}
    f(b)  # Runtime error: int + str

Rejected Alternatives
=====================

Several variations were considered and discarded:

* A ``readonly`` parameter to ``TypedDict``, behaving much like ``TypedMapping`` but with the additional constraint that instances must be dictionaries at runtime. This was discarded as less flexible due to the extra constraint; additionally, the new type nicely mirrors the existing ``Mapping``/``Dict`` types.
* Inheriting from a ``TypedMapping`` subclass and ``TypedDict`` resulting in mutator methods being added for all fields, not just those actively (re)declared in the class body. Discarded as less flexible, and not matching how inheritance works in other cases for ``TypedDict`` (e.g. total=False and total=True do not affect fields not specified in the class body).
* A generic type that removes mutator methods from its parameter, e.g. ``Readonly[MovieRecord]``. This would naturally want to be defined for a wider set of types than just ``TypedDict`` subclasses, and also raises questions about whether and how it applies to nested types. We decided to keep the scope of this PEP narrower.

> * Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just ``TypedMapping`` superclasses, no if there's a ``Protocol``).
> 

Reviewer: ```suggestion
* Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just ``TypedMapping`` superclasses, no if there's a ``Protocol``).

Copyright
=========
This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.
```

Update suggestion adding required copyright line after other changes.

PEP: 705
Title: TypedMapping: Type Hints for Mappings with a Fixed Set of Keys
Author: Alice Purcell <alicederyn@gmail.com>
Sponsor: Pablo Galindo <pablogsal@gmail.com>

> Discussions-To: pending

Author: Didn't realise this would block linting so much. Removing until we have a thread to link to.
```suggestion
```

Reviewer: See https://github.com/alicederyn/peps/pull/4#discussion_r1102348723 ; it makes that one check go red as expected, but doesn't interfere with any others.

> Discussions-To: https://mail.python.org/archives/list/python-dev@python.org/thread/2P26R4VH2ZCNNNOQCBZWEM4RNF35OXOW/

Reviewer 2: The discussions-to thread should be a new thread specifically about the PEP. Since this is very much a type system feature, I would strongly encourage you to have a discussion on typing-sig first.

Reviewer: And just to clarify, you can wait to create this thread until right before or right after this PR is merged—just make sure to update `Discussions-To` and add an entry to `Post-History` accordingly.

Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 07-Nov-2022
Python-Version: 3.12

> Post-History: `30-Sep-2022 <https://mail.python.org/archives/list/typing-sig@python.org/thread/6FR6RKNUZU4UY6B6RXC2H4IAHKBU3UKV/>`__,

Reviewer 2: Technically I don't think "pre-discussions" should be in this list, though they do provide useful context.

Reviewer: Well, as we've shifted the purpose of this field in modern PEPs toward documenting/linking the history of the PEP's discussion as to opposed to just the dates of its posting, it seems well within the spirit of the current guidance to include it (having written the current letter of it), so long as the topic of the discussion is the idea proposed in the PEP (which it is).

And I agree, it's quite useful. It also helps make it clear that the PEP has fulfilled the requirement:

> [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate

And it seems what many authors are already doing as of late, which has been helpful to me as a PEP reader and editor.

Reviewer: Marking this resolved for now as it seems like this is fine and the PEP is otherwise almost ready to merge

Motivation
==========

Representing structured data using (potentially nested) dictionaries with string keys is a common pattern in Python programs. :pep:`589` allows these values to be type checked when the exact type is known up-front, but it is hard to write read-only code code that accepts more specific variants: for instance, where fields may be subtypes or restrict a union of possible types. This is an especially common issue when writing APIs for services, which may support a wide range of input structures, and typically do not need to modify their input.

Take for example the function ``movie_string`` defined above. Suppose its type signature is defined using a ``TypedDict``::

    from typing import NotRequired, TypedDict

    class Movie(TypedDict):
        name: str
        year: NotRequired[int | None]

    class MovieRecord(TypedDict):
        name: str
        year: int

    def movie_string(movie: Movie) -> str:
        if movie.get("year") is None:
            return movie["name"]
        else:
            return f'{movie["name"]} ({movie["year"]})'

> Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::

Reviewer: ```suggestion
Attempting to pass a ``MovieRecord`` into ``movie_string`` results in the error (using mypy):

.. code-block:: text
```

* Fix grammar
* Use appropriate lexer for code block, instead of trying to parse it as Python by default

    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"

This particular use case should be type-safe, but the type checker correctly stops the

> user from passing a ``MovieRecord`` into a ``Movie`` argument in the general case because

Reviewer: ```suggestion
user from passing a ``MovieRecord`` into a ``Movie`` parameter in the general case, because
```

* Really needs a comma here to separate these clauses and avoid a runon
* Isn't the user passing a ``MovieRecord``-annotated _argument_ into a ``Move``-annotated _parameter_ here?

the ``Movie`` class has mutator methods that could potentially allow the function to break

> the type constraints in ``MovieRecord`` by setting ``movie["year"] = None`` or ``del movie["year"]``.

Reviewer: ```suggestion
the type constraints in ``MovieRecord`` (e.g. by ``movie["year"] = None`` or ``del movie["year"]``).
```

Further break up this run-on sentence, and clearly delineate that these are examples.

The problem disappears if we don't have mutator methods in ``Movie``. This could be achieved by defining an immutable interface using a :pep:`544` Protocol::

    from typing import Literal, Protocol, overload

    class Movie(Protocol):
        @overload
        def get(self, key: Literal["name"]) -> str: ...

        @overload
        def get(self, key: Literal["year"]) -> int | None: ...

        @overload
        def __getitem__(self, key: Literal["name"]) -> str: ...

        @overload
        def __getitem__(self, key: Literal["year"]) -> int | None: ...

This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.

> The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above.

Reviewer: ```suggestion
Rationale
======

The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and supports the cases exemplified above.
```

* IMO, this is really your Rationale section here, since it is explaining the benefits for your proposed _solution_, as opposed to laying out the _problem_,.
* Fix grammar (singular/plural agreement)

Also, it would be really nice to avoid overloading "type"/"types" so much to refer to three different things here, but I'm not sure exactly what to suggest.

In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.

Specification
=============

The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:

1. The runtime type of a TypedMapping object is not constrained to be a ``dict``

> 2. no mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated

Reviewer 2: We should define exactly what `TypedMapping` does support, not just what it doesn't. For example, does it support the `|` method, which is on `dict` but not `collections.abc.Mapping`?

Reviewer: Seems like this was addressed now, @JelleZijlstra ?

> 1. The runtime type of a TypedMapping object is not constrained to be a ``dict``
> 2. no mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
> 3. a class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
> 4. subclasses can narrow field types, in the same manner as other protocols

Reviewer: ```suggestion
* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``
* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
* Subclasses can narrow field types, in the same manner as other protocols
```

* Use a bulleted rather than ordered list for items that aren't sequential and don't have a required, unambiguous order
* Use consistent and correct capitalization
* Add missing literal to one ``TypedMapping`` instance

> All current and future features of TypedDict are applicable to TypedMapping, including class-based and alternative syntax, totality, and ``Required`` and ``NotRequired`` from :pep:`655`.

Reviewer: ```suggestion
All current and future features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

* Avoid potentially confusing/difficult to read double "and"
* Be consistent about using literals, at least within the same section

> All current and future features of TypedDict are applicable to TypedMapping, including class-based and alternative syntax, totality, and ``Required`` and ``NotRequired`` from :pep:`655`.

Reviewer 2: I would remove "and future". We don't know what future features will be added to TypedDict, and some may not apply to immutable mappings.

Reviewer: This bothered me too, as a non-typing expert.

Perhaps also a mention that any future proposed `TypedDict` features should explicitly state whether they are also supported on `TypedMapping` or not?

Reviewer: Added [a suggestion](https://github.com/python/peps/pull/2997#discussion_r1100916542) for your comment above

Reviewer: Seems like this comment was ultimately addressed by replacing this line with a more detailed enumeration of the specific features supported that didn't refer to future features.

> As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving TypedMapping and TypedDict objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.

Reviewer: ```suggestion
As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving ``TypedMapping`` and ``TypedDict`` objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.
```

* Be consistent about using literals or not, at least within the same section

> As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving TypedMapping and TypedDict objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.

Reviewer 2: Are there concrete reasons why we can't provide a full specification?

Author: Because it's incredibly complex and full of edge cases. For instance, type checkers are still working out how `@final` and `TypedDict` should interact (https://github.com/python/mypy/pull/13838). I think we'd take forever trying to nail a full spec down, and still get it wrong :(

Reviewer: @JelleZijlstra ?

Reviewer 2: Not entirely convinced. Type narrowing is indeed complicated, but that's different from providing a full specification of how type checking itself should work. In any case, this can be part of the conversation after the PEP is submitted.

Multiple inheritance and TypedDict
----------------------------------

A type that inherits from a TypedMapping protocol and from TypedDict (either directly or indirectly):

> 5. is the structural intersection of its parents, or invalid if no such intersection exists
> 6. instances must be a dict subclass
> 7. adds mutator methods only for fields it explicitly (re)declares

Reviewer: ```suggestion
* is the structural intersection of its parents, or invalid if no such intersection exists
* instances must be a dict subclass
* adds mutator methods only for fields it explicitly (re)declares
```

I was very confused about the numbering here, as not only is there no unambiguous declared sequence, but it does not appear to be a direct continuation of the previous numbered list, nor was there another obvious reasons for this.

For example::

    class Movie(TypedMapping):
        name: str
        year: int | None

>     class MovieRecord(A, TypedDict):

Reviewer: ```suggestion
    class MovieRecord(Movie, TypedDict):
```

Do you mean ``Movie`` here, to match with the description above? ``A`` isn't defined anywhere up to this point.

        year: int

    movie: MovieRecord = { "name": "Blade Runner",
                           "year": 1982 }
     

>     movie["year"] = 1985  # Fine, mutator methods added in definition
>     movie["name"] = "Terminator"  # Type check error, "name" mutator not declared

Reviewer: ```suggestion
    movie["year"] = 1985  # Fine; mutator methods added in definition
    movie["name"] = "Terminator"  # Type check error; "name" mutator not declared
```

Fix grammar errors (comma splices)

Inheriting, directly or indirectly, from both TypedDict and Protocol will continue to fail at runtime, and should continue to be rejected by type checkers.

Multiple inheritance and Protocol
---------------------------------

* A type that inherits from a TypedMapping protocol and from a Protocol protocol must satisfy the protocols defined by both, but is not itself a protocol unless it inherits directly from TypedMapping or Protocol.
* A type that inherits from a TypedMapping protocol and from Protocol itself is configured as a Protocol. Methods and properties may be defined; keys may not::

    class A(Movie, Protocol):
        # Declare a mutable property called 'year'
        # This does not affect the dictionary key 'year'
        year: str

* A type that inherits from a Protocol protocol and from TypedMapping itself is configured as a TypedMapping. Keys may be defined; methods and properties may not::

    class B(A, TypedMapping):
        # Declare a key 'year'
        # This does not affect the property 'year'
        year: int

> Type Consistency Rules

Reviewer: ```suggestion
Type consistency rules
```

Either title case or sentence case is fine for non-top-level headings, so long as you pick one and are (relatively) consistent about it. My suggestion changes this one, but you could equally well change the previous two, if you preferred.

    b: B = {'x': 0, 'y': 'foo'}
    f(b)  # Runtime error: int + str

Rejected Alternatives
=====================

Several variations were considered and discarded:

* A ``readonly`` parameter to ``TypedDict``, behaving much like TypedMapping but with the additional constraint that instances must be dictionaries at runtime. This was discarded as less flexible due to the extra constraint; additionally, the new type nicely mirrors the existing ``Mapping``/``Dict`` types.
* Inheriting from a ``TypedMapping`` subclass and ``TypedDict`` resulting in mutator methods being added for all fields, not just those actively (re)declared in the class body. Discarded as less flexible, and not matching how inheritance works in other cases for TypedDict (e.g. total=False and total=True do not affect fields not specified in the class body).
* A generic type that removes mutator methods from its parameter, e.g. ``Readonly[MovieRecord]``. This would naturally want to be defined for a wider set of types than just ``TypedDict`` subclasses, and also raises questions about whether and how it applies to nested types. We decided to keep the scope of this PEP narrower.

> * Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just TypedMappings, no if there's a Protocol)
> 

Reviewer: ```suggestion
* Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just TypedMappings, no if there's a Protocol).

Copyright
=========

This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.
```

Add missing required copyright section. Also, fix missing period and trailing blank lines.

Reviewer: Superseded by [followup suggestion](https://github.com/python/peps/pull/2997#discussion_r1100915336) with just the remaining required copyright change

Rationale
=========

The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above::

    from typing import NotRequired, TypedMapping

    class Movie(TypedMapping):
        name: str
        year: NotRequired[int | None]

In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.

Specification
=============

A ``TypedMapping`` type defines a protocol with the same methods as :class:`~collections.abc.Mapping`, but with value types determined per-key as with ``TypedDict``.

Notable similarities to ``TypedDict``:

* A ``TypedMapping`` protocol can be declared using class-based or alternative syntax.
* Keys must be strings.

> * By default, all specified keys must be present in a ``TypedMapping`` instance. It is posible to override this by specifying totality, or by using ``NotRequired`` from :pep:`655`.

Reviewer 3: Found via `make spellcheck`:

```suggestion
* By default, all specified keys must be present in a ``TypedMapping`` instance. It is possible to override this by specifying totality, or by using ``NotRequired`` from :pep:`655`.
```

    g(c)  # Runtime error: str + int

* A ``TypedMapping`` with all ``int`` values is not consistent with ``Mapping[str, int]``, since there may be additional non-``int`` values not visible through the type, due to structural subtyping. This mirrors ``TypedDict``. Example::

    class A(TypedMapping):
        x: int

    class B(TypedMapping):
        x: int
        y: str

    def sum_values(m: Mapping[str, int]) -> int:
        return sum(m.values())

    def f(a: A) -> None:
        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]

    b: B = {'x': 0, 'y': 'foo'}
    f(b)  # Runtime error: int + str

Backwards Compatibility
=======================

This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
inherit from ``TypedMapping`` protocols in a way that changes the resulting
overloads), so code that inspects ``TypedDict`` types will have to change. This
is expected to mainly affect type-checkers.

The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
enabling its use in older versions of Python. Users wanting to inherit from
both ``TypedDict`` and ``TypedMapping`` will need to use

> ``typing_extensions.TypedDict`` in these versions, as the ``typing`` declaration

Reviewer: ```suggestion
``typing_extensions.TypedDict`` in these versions, as the existing ``typing`` implementation
```

* "implementation" is a much more appropriate word than "declaration" here
* Clarify it is the ``typing`` implementation in _existing_ versions but not going forward

> of ``TypedDict`` does an ``is`` check on the metaclass of all base classes.

Reviewer: I would at least briefly explicitly mention why this is a problem.

Author: ```suggestion
of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
an error if any other metaclass is found. This would cause any subclass of both
this implementation and a ``TypedMapping`` protocol to trigger an error at runtime.
```
How about this?

Reviewer: Looks great, thanks! :+1:

    class A(TypedMapping):
        x: int

    class B(TypedMapping):
        x: int
        y: str

    def sum_values(m: Mapping[str, int]) -> int:
        return sum(m.values())

    def f(a: A) -> None:
        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]

    b: B = {'x': 0, 'y': 'foo'}
    f(b)  # Runtime error: int + str

Backwards Compatibility
=======================

This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
inherit from ``TypedMapping`` protocols in a way that changes the resulting
overloads), so code that inspects ``TypedDict`` types will have to change. This
is expected to mainly affect type-checkers.

The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
enabling its use in older versions of Python. Users wanting to inherit from
both ``TypedDict`` and ``TypedMapping`` will need to use
``typing_extensions.TypedDict`` in these versions, as the existing ``typing`` implementation
of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
an error if any other metaclass is found. This would cause any subclass of both

> that implementation and a ``TypedMapping`` protocol to trigger an error at runtime.
> 
> 
> How to Teach This

Reviewer: ```suggestion
that implementation and a ``TypedMapping`` protocol to trigger an error at runtime.

Security Implications
=====================

There are no known security consequences arising from this PEP.

How to Teach This
```

If there are no security implications, then the general convention is to just explicitly state this for clarity.

Author: Ah, I looked at the recently-accepted https://peps.python.org/pep-0702/ which doesn't have this section.

Reviewer: Yeah, we've gotten gradually more thorough lately for new PEPs, especially with the new checklist. It's not a hard requirement if there's a specific reason a section clearly isn't relevant, but it doesn't hurt to be explicit.

=================

> This PEP should be linked in the ``typing`` module's documentation.

Reviewer: Generally speaking, PEPs are primarily change proposals and relying on them as the main living, user-friendly documentation for a feature should really be avoided. However, in this case this is mostly up to the typing community, which I understand is working on a plan to have a canonical specs and/or docs site like the packaging community does, and can be decided and refined later if and when the PEP is accepted.

Author: I used the recently-accepted https://peps.python.org/pep-0692/ as my model here. The typing module documentation does seem to have a lot of links to PEPs. Would definitely be a fan of that changing.

> This PEP should be linked in the ``typing`` module's documentation.

Reviewer: In any case, link the module too:

```suggestion
This PEP should be linked in the :mod:`typing` module's documentation.
```

    g(c)  # Runtime error: str + int

* A ``TypedMapping`` with all ``int`` values is not consistent with ``Mapping[str, int]``, since there may be additional non-``int`` values not visible through the type, due to structural subtyping. This mirrors ``TypedDict``. Example::

    class A(TypedMapping):
        x: int

    class B(TypedMapping):
        x: int
        y: str

    def sum_values(m: Mapping[str, int]) -> int:
        return sum(m.values())

    def f(a: A) -> None:
        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]

    b: B = {'x': 0, 'y': 'foo'}
    f(b)  # Runtime error: int + str

Backwards Compatibility
=======================

This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
inherit from ``TypedMapping`` protocols in a way that changes the resulting
overloads), so code that inspects ``TypedDict`` types will have to change. This
is expected to mainly affect type-checkers.

The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
enabling its use in older versions of Python. Users wanting to inherit from
both ``TypedDict`` and ``TypedMapping`` will need to use

> ``typing_extensions.TypedDict`` in these versions, as the existing ``typing`` implementation

Reviewer 2: This is a bit too much implementation detail. We may well be able to work around this restriction with some creative work in typing_extensions. I would suggest taking out everything starting with "Users wanting to inherit". Such limitations can then be explained in the `typing_extensions` documentation.

of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
an error if any other metaclass is found. This would cause any subclass of both
that implementation and a ``TypedMapping`` protocol to trigger an error at runtime.

Security Implications
=====================

There are no known security consequences arising from this PEP.

How to Teach This
=================

> This PEP should be linked in the :mod:`typing` module's documentation.

Reviewer 2: It's true that we've historically often linked PEPs in the typing docs, but as mentioned above this isn't an ideal state. I'm hopeful that we'll eventually get to rewrite the docs in a way that doesn't require linking to PEPs, so I don't think this PEP should mandate that the typing docs add this specific link.

Author: Changed to no longer be mandatory, but left as a suggestion to mirror https://peps.python.org/pep-0692/#how-to-teach-this &mdash; do let me know if you'd rather leave it out entirely.

Class documentation should also be added, using that for
:class:`~collections.abc.Mapping`, :class:`~typing.Protocol` and
:class:`~typing.TypedDict` as examples.

Suggested introductory sentence: "Base class for read-only mapping protocol classes."

Reference Implementation
========================

> No reference implementation exists yet.

Reviewer 2: I would encourage you to submit a typing-extensions implementation as soon as possible.

Author: Understood, thanks.
